v0.14.48 panic: runtime error: index out of range

Syncthing Version: syncthing v0.14.48 "Dysprosium Dragonfly" (go1.10.2 linux-386) teamcity@build.syncthing.net 2018-05-14 06:53:06 UTC
OS Version: Slackware 14.2 32bit


Panics at irregular intervals; sometimes as soon as ~40s after startup, sometimes several hours.  Always the message is "panic: runtime error: index out of range".  
Presumably my machine acquired v0.14.48 within twelve hours of its release, which would put the first panic 24hrs or more since release.  

From the first log file:
[7JAPU] 16:16:31 INFO: Folder DABingeNeedRetire isn't making progress. Pausing puller for 1m0s.
Panic at 2018-06-08T16:17:16-08:00
panic: runtime error: index out of range

goroutine 247254 [running]:
[....]


Switching back to v0.14.47 (And turning off auto update in the config) fixes this problem.

Thanks.

```
Panic at 2018-06-10T16:12:50-08:00
panic: runtime error: index out of range

goroutine 673 [running]:
github.com/syncthing/syncthing/lib/model.(*Model).recheckFile(0x194d2420, 0xc326c38c, 0x79b45524, 0x9b71beeb, 0x4b0f3c16, 0x53a919b2, 0xe2119819, 0xc90f78a8, 0x112cf36c, 0x870fde0, ...)
	/opt/buildagent/work/src/github.com/syncthing/syncthing/lib/model/model.go:1410 +0xfd6
github.com/syncthing/syncthing/lib/model.(*Model).Request(0x194d2420, 0xc326c38c, 0x79b45524, 0x9b71beeb, 0x4b0f3c16, 0x53a919b2, 0xe2119819, 0xc90f78a8, 0x112cf36c, 0x1a6a9360, ...)
	/opt/buildagent/work/src/github.com/syncthing/syncthing/lib/model/model.go:1386 +0x10b0
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).handleRequest(0x1a54ad80, 0x8, 0x1a6a9360, 0xb, 0x1b03f2c0, 0x2d, 0x80000, 0x0, 0x181e8, 0x19570520, ...)
	/opt/buildagent/work/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:588 +0x148
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop
	/opt/buildagent/work/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:380 +0x586

```

--- a/lib/model/model.go
+++ b/lib/model/model.go
@@ -1402,7 +1402,7 @@ func (m *Model) recheckFile(deviceID protocol.DeviceID, folderFs fs.Filesystem, 
 		return
 	}
 
-	if blockIndex > len(cf.Blocks) {
+	if blockIndex >= len(cf.Blocks) {
 		l.Debugf("%v recheckFile: %s: %q / %q i=%d: block index too far", m, deviceID, folder, name, blockIndex)
 		return
 	}

--- a/lib/model/model_test.go
+++ b/lib/model/model_test.go
@@ -3611,7 +3611,7 @@ func TestIssue4903(t *testing.T) {
 }
 
 func TestIssue5002(t *testing.T) {
-	// recheckFile should not panic when given an index equal to the number of blocks plus one
+	// recheckFile should not panic when given an index equal to the number of blocks
 
 	db := db.OpenMemory()
 	m := NewModel(defaultCfgWrapper, protocol.LocalDeviceID, "syncthing", "dev", db, nil)

--- a/lib/model/model_test.go
+++ b/lib/model/model_test.go
@@ -3610,6 +3610,32 @@ func TestIssue4903(t *testing.T) {
 	}
 }
 
+func TestIssue5002(t *testing.T) {
+	// recheckFile should not panic when given an index equal to the number of blocks plus one
+
+	db := db.OpenMemory()
+	m := NewModel(defaultCfgWrapper, protocol.LocalDeviceID, "syncthing", "dev", db, nil)
+	m.AddFolder(defaultFolderConfig)
+	m.StartFolder("default")
+
+	m.ServeBackground()
+	defer m.Stop()
+
+	if err := m.ScanFolder("default"); err != nil {
+		t.Error(err)
+	}
+
+	file, ok := m.CurrentFolderFile("default", "foo")
+	if !ok {
+		t.Fatal("test file should exist")
+	}
+	nBlocks := len(file.Blocks)
+
+	m.recheckFile(protocol.LocalDeviceID, defaultFolderConfig.Filesystem(), "default", "foo", nBlocks-1, []byte{1, 2, 3, 4})
+	m.recheckFile(protocol.LocalDeviceID, defaultFolderConfig.Filesystem(), "default", "foo", nBlocks, []byte{1, 2, 3, 4}) // panic
+	m.recheckFile(protocol.LocalDeviceID, defaultFolderConfig.Filesystem(), "default", "foo", nBlocks+1, []byte{1, 2, 3, 4})
+}
+
 func addFakeConn(m *Model, dev protocol.DeviceID) *fakeConnection {
 	fc := &fakeConnection{id: dev, model: m}
 	m.AddConnection(fc, protocol.HelloResult{})

